designs/plugin-resolution.md

```diff
@@ -0,0 +1,88 @@
+# Plugin and project version resolution
+
+## Old behavior
+
+The same process is followed for all commands, without distinction between `init` and the rest. The only difference is
+that, for `init`, the configuration file will not be present.
+
+1. Obtain the project version from project configuration file's `version` field, `--project-verison` flag, or default
+   project version configured during CLI creation (`cli.New(cli.WithDefaultProjectVersion(...))`).
+    * In case both a project configuration file and the flag are found, fail if they are different.
+2. Obtain the plugin keys from the project configuration file's `layout` field, `--plugins` flag, or default plugins
+   configured during CLI creation for the above project version (`cli.New(cli.WithDefaultPlugins(...))`).
+    * In case both a project configuration file and the flag are found, fail if they are different.
+3. In case any of the plugin keys is not fully qualified (full name and version), check if there is only one plugin that
+   fits that unqualified plugin key.
+4. Resolve the plugins with the list of qualified keys above.
+5. Verify the plugins support the project version.
+   
+
+### Flaws
+
+* Default plugins need to be provided per project version
+* Project version can't be resolved (e.g., when a single project version is supported by the provided plugins).
+* Command calls are not allowed to override plugin keys.
+
+## Proposal
+
+The following plugin resolution algorithm solves the flaws described above:
+
+1. [All commands but Init] Obtain the project version and plugin keys from the project configuration if available.
+     * If they are available, plugin keys will be fully qualified.
+2. [Init command only] Obtain the project version from `--project-version` flag.
+     * The flag is optional, we will try to resolve project version later in case it is not provided.
+3. Obtain the optional plugin keys from `--plugins` flag.
+     * [Init command only] If not present, use default plugins.
+     * [All commands but Init] If not present, use plugin keys from configuration file (step 1).
+4. Qualify the unqualified keys from step 3.
+     * If we know the project version, use this info to filter the available plugins so that we can be more accurate.
+5. Resolve the plugins with the list of qualified keys above.
+6. [Init command only] Resolve the project version in case it wasn't provided and there is only one project version
+   supported by all the resolved plugins.
+   
```

Comment:
Should we not cover here the subcommand plugins scenarios when the layout ( init ) has more than one plugin in this proposal?

Reply:
From a plugin resolution perspective, there is no difference. We are already resolving multiples plugins. It is latter when we are forced to only have 1. This EP only talks about the resolution proposal.